fdbserver/TLogServer.actor.cpp

```diff
@@ -336,7 +336,7 @@ struct TLogData : NonCopyable {
 	int64_t overheadBytesInput;
 	int64_t overheadBytesDurable;
 	int activePeekStreams = 0;
-
+	Optional<Version> clusterRecoveryVersion;
```

Comment:
what if there are multiple recoveries? Say, a tlog crashed, and triggered the first recovery. Then a second recovery happened. At this time, the crashed tlog comes back. What recovery version will this tlog gets, first recovery or second recovery?

Reply:
When a new epoch starts, we will lock tLogs from the previous generation and send a message to each with the RV. 

In the first recovery RV1 would be sent, but in your example the tlogX had crashed, so it would not be sent any message. We send the RV to the set of tLogs we can lock from the previous generation.

In the second recovery the tLogX comes back. It could receive the message with RV2. But tLogX is not a member of the set of locked tLogs, as it is not a member of the previous generation. it would be in the structure `oldLogData`. We should ensure members of `oldLogData` are sent the message as well. 

I will think about how to test it the scenario.

Reply:
I believe we set recoverAt whenever a new epoch is created [here](https://github.com/apple/foundationdb/blob/46e985acf348d351205f80d799321f826fffcd91/fdbserver/TagPartitionedLogSystem.actor.cpp#L2926). Old epoch's are stored. In your example, we would use the first recovery's value.